### PR TITLE
Build rules_kotlin using Kotlin 1.6

### PR DESCRIPTION
**Pull request by Bencodes**
This PR also includes:
- Updating Dagger for the entire project, including the example apps
- Updating Anvil in the example apps
- Removed `-Xuse-experimental=kotlin.Experimental` as `-Xopt-in` is now the default for 1.6.x
- Building rules_kotlin against the latest 1.6.20 release
- Making Kotlin 1.6 the default version

https://github.com/JetBrains/kotlin/releases/tag/v1.6.20

### CODE REVIEW

**Comment by Bencodes** [comment]
Will wait for https://github.com/bazelbuild/rules_kotlin/pull/625/files to merge

**Comment by Bencodes** [comment]
1.6.10-RC is out with some critical fixes that we need. Going to wait until that rolls out to merge this.

**Comment by nkoroste** [comment]
> 1.6.10-RC is out with some critical fixes that we need. Going to wait until that rolls out to merge this.

hopefully they land https://github.com/JetBrains/kotlin/pull/4677 in there as well

**Comment by sfs** [comment]
> hopefully they land JetBrains/kotlin#4677 in there as well

Sorry, the PR was uploaded after the RC release and it didn't make it into 1.6.10. Does this issue affect you? I was hoping that nobody else was using ABI jars as dependencies in IntelliJ.

**Review by nkoroste — APPROVED** [review]

**Comment by Bencodes** [comment]
@sfs any idea when 1.6.20 will hit RC so that we can do another round of testing?

File: src/test/kotlin/io/bazel/kotlin/builder/tasks/jvm/KotlinBuilderJvmAbiTest.java
Comment (Bencodes) [inline src/test/kotlin/io/bazel/kotlin/builder/tasks/jvm/KotlinBuilderJvmAbiTest.java:41]: Without this, the skip-code-gen plugin gets added, which causes the new ABI compiler to short circuit and not output anything.

**Comment by Bencodes** [comment]
This is still blocked by the metadata fixes that will go out in `1.6.20`. We can work around this if we need to by embedding the patched ABI gen compiler in here if needed.

**Comment by brettwooldridge** [comment]
@Bencodes Kotlin v1.6.20 is out now.

**Comment by Bencodes** [comment]
@restingbull @nkoroste this should be good to merge now if you want to give it a once over.

**Comment by Bencodes** [comment]
I'll update the compose example after they update that with the latest Kotlin compatibility.

**Review by restingbull — APPROVED** [review]

**Pull request merged**
